src/Components/QuickGrid/Microsoft.AspNetCore.Components.QuickGrid/src/QuickGrid.razor.cs

```diff
@@ -0,0 +1,432 @@
+    /// (either <see cref="Items"/> or <see cref="ItemsProvider"/>).
+    /// </summary>
+    /// <returns>A <see cref="Task"/> that represents the completion of the operation.</returns>
+    public async Task RefreshDataAsync()
+    {
+        await RefreshDataCoreAsync();
+        StateHasChanged();
+    }
+
+    // Same as RefreshDataAsync, except without forcing a re-render. We use this from OnParametersSetAsync
+    // because in that case there's going to be a re-render anyway.
+    private async Task RefreshDataCoreAsync()
+    {
+        // Move into a "loading" state, cancelling any earlier-but-still-pending load
+        _pendingDataLoadCancellationTokenSource?.Cancel();
+        var thisLoadCts = _pendingDataLoadCancellationTokenSource = new CancellationTokenSource();
+
+        if (_virtualizeComponent is not null)
+        {
+            // If we're using Virtualize, we have to go through its RefreshDataAsync API otherwise:
+            // (1) It won't know to update its own internal state if the provider output has changed
+            // (2) We won't know what slice of data to query for
+            await _virtualizeComponent.RefreshDataAsync();
+            _pendingDataLoadCancellationTokenSource = null;
+        }
+        else
+        {
+            // If we're not using Virtualize, we build and execute a request against the items provider directly
+            _lastRefreshedPaginationStateHash = Pagination?.GetHashCode();
+            var startIndex = Pagination is null ? 0 : (Pagination.CurrentPageIndex * Pagination.ItemsPerPage);
+            var request = new GridItemsProviderRequest<TGridItem>(
+                startIndex, Pagination?.ItemsPerPage, _sortByColumn, _sortByAscending, thisLoadCts.Token);
+            var result = await ResolveItemsRequestAsync(request);
+            if (!thisLoadCts.IsCancellationRequested)
+            {
+                _currentNonVirtualizedViewItems = result.Items;
+                _ariaBodyRowCount = _currentNonVirtualizedViewItems.Count;
+                Pagination?.SetTotalItemCountAsync(result.TotalItemCount);
+                _pendingDataLoadCancellationTokenSource = null;
+            }
+        }
+    }
+
+    // Gets called both by RefreshDataCoreAsync and directly by the Virtualize child component during scrolling
+    private async ValueTask<ItemsProviderResult<(int, TGridItem)>> ProvideVirtualizedItems(ItemsProviderRequest request)
+    {
+        _lastRefreshedPaginationStateHash = Pagination?.GetHashCode();
+
+        // Debounce the requests. This eliminates a lot of redundant queries at the cost of slight lag after interactions.
+        // TODO: Consider making this configurable, or smarter (e.g., doesn't delay on first call in a batch, then the amount
+        // of delay increases if you rapidly issue repeated requests, such as when scrolling a long way)
+        await Task.Delay(100);
```

Comment:
Is now the time to make this configurable?

Reply:
I don't think so. I feel like there's a few ways to pursue this particular TODO that would require some time to figure out. I think I'll take the small handful of remaining TODOs in here and make small tasks on GitHub that community members could grab - or us depending on capacity.